src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs

```diff
@@ -55,57 +61,196 @@ internal static partial class JsonWriterHelper
+
+                        // Check if any of the 16 bytes need to be escaped.
+                        Vector128<sbyte> mask = CreateEscapingMask(sourceValue);
+
+                        int index = Sse2.MoveMask(mask.AsByte());
+                        // If index == 0, that means none of the 16 bytes needed to be escaped.
+                        // TrailingZeroCount is relatively expensive, avoid it if possible.
+                        if (index != 0)
+                        {
+                            // Found at least one byte that needs to be escaped, figure out the index of
+                            // the first one found that needed to be escaped within the 16 bytes.
+                            idx += BitOperations.TrailingZeroCount(index | 0xFFFF0000);
+                            goto Return;
+                        }
+                        idx += 16;
+                        startingAddress += 16;
+                    }
+
+                    // Process the remaining characters.
+                    Debug.Assert(value.Length - idx < 16);
+                }
+#endif
+
+                for (; idx < value.Length; idx++)
+                {
+                    Debug.Assert((ptr + idx) <= (ptr + value.Length));
+                    if (NeedsEscaping(*(ptr + idx)))
+                    {
+                        goto Return;
+                    }
+                }
 
-        Return:
-            return idx;
+                idx = -1; // all characters allowed
+
+            Return:
+                return idx;
+            }
         }
 
         public static unsafe int NeedsEscaping(ReadOnlySpan<char> value, JavaScriptEncoder encoder)
         {
-            int idx;
-
-            // Some implementations of JavascriptEncoder.FindFirstCharacterToEncode may not accept
-            // null pointers and gaurd against that. Hence, check up-front and fall down to return -1.
-            if (encoder != null && !value.IsEmpty)
+            fixed (char* ptr = value)
             {
-                fixed (char* ptr = value)
+                int idx = 0;
+
+                // Some implementations of JavascriptEncoder.FindFirstCharacterToEncode may not accept
+                // null pointers and gaurd against that. Hence, check up-front and fall down to return -1.
+                if (encoder != null && !value.IsEmpty)
                 {
                     idx = encoder.FindFirstCharacterToEncode(ptr, value.Length);
+                    goto Return;
                 }
-                goto Return;
-            }
 
-            for (idx = 0; idx < value.Length; idx++)
-            {
-                if (NeedsEscaping(value[idx]))
+#if BUILDING_INBOX_LIBRARY
+                if (Sse2.IsSupported)
                 {
-                    goto Return;
+                    short* startingAddress = (short*)ptr;
+                    while (value.Length - 8 >= idx)
```

Comment:
For `value.Length >= 16` did you consider packing into a byte-vector (with saturation), so 16 elements at once can be scanned?

Sse2 implies little-endian, so there should be no problems for that (endianess-wise).
The code for the byte-path can be reused (needs to be extracted to a method, so the loops remain independent).

Outline:
#### elements >= 16 

pack two char-vectors to on byte-vector, check as bytes, transform back idx

#### elements >= 8 and the remainder from above

pack the one char-vector with a dummy char-vector to one byte-vector, same as above.
For the dummy char-vector use one that is known to not need escaping, e.g. `Vector128.Create((short)'A')`.

#### elements < 8 and the remainder from above

sequential scan

Reply:
Tried this in https://github.com/gfoidl/corefx/commit/3b1922c2cab1bd60a89a72e2e15b0f11d4645a64

For small input the overhead is too high, for larger inputs it's faster.

<details>
  <summary>benchmark results</summary>

```
|                 Method | TestStringData |     Mean |     Error |    StdDev | Ratio | RatioSD |
|----------------------- |--------------- |---------:|----------:|----------:|------:|--------:|
| NeedsEscapingCurrentPR |          1, -1 | 10.26 ns | 0.0481 ns | 0.0427 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |          1, -1 | 10.97 ns | 0.0455 ns | 0.0403 ns |  1.07 |    0.00 |
|                        |                |          |           |           |       |         |
| NeedsEscapingCurrentPR |           1, 0 | 10.30 ns | 0.0667 ns | 0.0592 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |           1, 0 | 10.79 ns | 0.2245 ns | 0.2100 ns |  1.04 |    0.02 |
|                        |                |          |           |           |       |         |
| NeedsEscapingCurrentPR |        100, -1 | 77.95 ns | 0.5453 ns | 0.5101 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |        100, -1 | 28.28 ns | 0.1341 ns | 0.1120 ns |  0.36 |    0.00 |
|                        |                |          |           |           |       |         |
| NeedsEscapingCurrentPR |         15, -1 | 21.73 ns | 0.2506 ns | 0.2344 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |         15, -1 | 20.35 ns | 0.1453 ns | 0.1288 ns |  0.94 |    0.01 |
|                        |                |          |           |           |       |         |
| NeedsEscapingCurrentPR |         16, -1 | 17.94 ns | 0.1079 ns | 0.0957 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |         16, -1 | 11.70 ns | 0.0390 ns | 0.0365 ns |  0.65 |    0.00 |
|                        |                |          |           |           |       |         |
| NeedsEscapingCurrentPR |         17, -1 | 19.26 ns | 0.1605 ns | 0.1502 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |         17, -1 | 13.18 ns | 0.0340 ns | 0.0284 ns |  0.68 |    0.01 |
|                        |                |          |           |           |       |         |
| NeedsEscapingCurrentPR |          2, -1 | 11.43 ns | 0.0287 ns | 0.0269 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |          2, -1 | 12.18 ns | 0.1658 ns | 0.1551 ns |  1.07 |    0.01 |
|                        |                |          |           |           |       |         |
| NeedsEscapingCurrentPR |         32, -1 | 26.47 ns | 0.1092 ns | 0.0968 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |         32, -1 | 14.06 ns | 0.0918 ns | 0.0814 ns |  0.53 |    0.00 |
|                        |                |          |           |           |       |         |
| NeedsEscapingCurrentPR |          4, -1 | 13.70 ns | 0.0802 ns | 0.0711 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |          4, -1 | 14.33 ns | 0.0611 ns | 0.0542 ns |  1.05 |    0.01 |
|                        |                |          |           |           |       |         |
| NeedsEscapingCurrentPR |          7, -1 | 17.12 ns | 0.1120 ns | 0.1048 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |          7, -1 | 17.77 ns | 0.1273 ns | 0.1191 ns |  1.04 |    0.01 |
|                        |                |          |           |           |       |         |
| NeedsEscapingCurrentPR |           7, 0 | 10.26 ns | 0.0563 ns | 0.0499 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |           7, 0 | 10.62 ns | 0.0696 ns | 0.0617 ns |  1.04 |    0.01 |
|                        |                |          |           |           |       |         |
| NeedsEscapingCurrentPR |           7, 1 | 11.39 ns | 0.0642 ns | 0.0569 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |           7, 1 | 11.80 ns | 0.0403 ns | 0.0377 ns |  1.04 |    0.01 |
|                        |                |          |           |           |       |         |
| NeedsEscapingCurrentPR |           7, 2 | 12.57 ns | 0.0758 ns | 0.0709 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |           7, 2 | 12.85 ns | 0.0707 ns | 0.0627 ns |  1.02 |    0.00 |
|                        |                |          |           |           |       |         |
| NeedsEscapingCurrentPR |           7, 3 | 13.68 ns | 0.0965 ns | 0.0902 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |           7, 3 | 14.07 ns | 0.1104 ns | 0.1033 ns |  1.03 |    0.01 |
|                        |                |          |           |           |       |         |
| NeedsEscapingCurrentPR |           7, 4 | 14.80 ns | 0.0444 ns | 0.0394 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |           7, 4 | 15.39 ns | 0.2170 ns | 0.2030 ns |  1.04 |    0.01 |
|                        |                |          |           |           |       |         |
| NeedsEscapingCurrentPR |           7, 5 | 15.88 ns | 0.0575 ns | 0.0509 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |           7, 5 | 16.32 ns | 0.0760 ns | 0.0674 ns |  1.03 |    0.01 |
|                        |                |          |           |           |       |         |
| NeedsEscapingCurrentPR |           7, 6 | 17.07 ns | 0.1173 ns | 0.1097 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |           7, 6 | 17.41 ns | 0.0618 ns | 0.0516 ns |  1.02 |    0.01 |
|                        |                |          |           |           |       |         |
| NeedsEscapingCurrentPR |          8, -1 | 13.65 ns | 0.0636 ns | 0.0531 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |          8, -1 | 12.16 ns | 0.0460 ns | 0.0430 ns |  0.89 |    0.01 |
|                        |                |          |           |           |       |         |
| NeedsEscapingCurrentPR |           8, 0 | 13.70 ns | 0.1083 ns | 0.1013 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |           8, 0 | 12.02 ns | 0.0529 ns | 0.0495 ns |  0.88 |    0.01 |
|                        |                |          |           |           |       |         |
| NeedsEscapingCurrentPR |           8, 1 | 13.61 ns | 0.0503 ns | 0.0420 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |           8, 1 | 11.99 ns | 0.0531 ns | 0.0497 ns |  0.88 |    0.00 |
|                        |                |          |           |           |       |         |
| NeedsEscapingCurrentPR |           8, 2 | 13.66 ns | 0.0927 ns | 0.0867 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |           8, 2 | 12.07 ns | 0.0559 ns | 0.0467 ns |  0.88 |    0.01 |
|                        |                |          |           |           |       |         |
| NeedsEscapingCurrentPR |           8, 3 | 13.64 ns | 0.0661 ns | 0.0586 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |           8, 3 | 12.05 ns | 0.0525 ns | 0.0466 ns |  0.88 |    0.00 |
|                        |                |          |           |           |       |         |
| NeedsEscapingCurrentPR |           8, 4 | 13.86 ns | 0.2429 ns | 0.2272 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |           8, 4 | 12.43 ns | 0.1654 ns | 0.1547 ns |  0.90 |    0.02 |
|                        |                |          |           |           |       |         |
| NeedsEscapingCurrentPR |           8, 5 | 13.93 ns | 0.0942 ns | 0.0881 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |           8, 5 | 12.00 ns | 0.0999 ns | 0.0934 ns |  0.86 |    0.01 |
|                        |                |          |           |           |       |         |
| NeedsEscapingCurrentPR |           8, 6 | 13.67 ns | 0.0711 ns | 0.0665 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |           8, 6 | 12.06 ns | 0.0583 ns | 0.0517 ns |  0.88 |    0.01 |
|                        |                |          |           |           |       |         |
| NeedsEscapingCurrentPR |           8, 7 | 13.68 ns | 0.0547 ns | 0.0427 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |           8, 7 | 12.03 ns | 0.0366 ns | 0.0342 ns |  0.88 |    0.00 |
|                        |                |          |           |           |       |         |
| NeedsEscapingCurrentPR |          9, -1 | 14.87 ns | 0.1161 ns | 0.1029 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |          9, -1 | 13.54 ns | 0.0464 ns | 0.0411 ns |  0.91 |    0.01 |
```
</details>

Will try to minimize the overhead...

Reply:
https://github.com/gfoidl/corefx/commit/b68062ee424118455cf95854c709a945f4fb187c -- it won't get any better

```
|                 Method | TestStringData |     Mean |     Error |    StdDev | Ratio |
|----------------------- |--------------- |---------:|----------:|----------:|------:|
| NeedsEscapingCurrentPR |          1, -1 | 10.13 ns | 0.0801 ns | 0.0750 ns |  1.00 |
|   NeedsEscapingBitMask |          1, -1 | 10.62 ns | 0.0289 ns | 0.0226 ns |  1.05 |
|                        |                |          |           |           |       |
| NeedsEscapingCurrentPR |           1, 0 | 10.24 ns | 0.0296 ns | 0.0263 ns |  1.00 |
|   NeedsEscapingBitMask |           1, 0 | 10.15 ns | 0.0511 ns | 0.0453 ns |  0.99 |
|                        |                |          |           |           |       |
| NeedsEscapingCurrentPR |          2, -1 | 11.44 ns | 0.0473 ns | 0.0419 ns |  1.00 |
|   NeedsEscapingBitMask |          2, -1 | 11.66 ns | 0.0330 ns | 0.0275 ns |  1.02 |
```

For me that would be acceptable.

--- 

PS: sorry for hijacking your PR :wink: (for today I'm finished...)

Reply:
The results here look promising. Please feel free to submit a follow up PR with the changes you want to explore and we can discuss the additional improvements there, along with the Ssse3 based implementation below :)

Running the whole set of benchmarks that I shared in the gist would be great, especially the serializer ones

Reply:
I'll prepare a PR for this. Thanks for the json-file.

Reply:
@ahsonkhan for the serializer-benchmarks I have to follow https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-corefx.md?

(FYI: code for PR is ready, benchmarks are outstanding -- the next days I may be out of office, so it can take some days until everything is ready to go)